packages/cli/src/services/orchestration/main/handle-workflow-updated.ts

```diff
@@ -0,0 +1,73 @@
+import { ActiveWorkflowRunner } from '@/ActiveWorkflowRunner';
+import { WorkflowRepository } from '@/databases/repositories/workflow.repository';
+import { Push } from '@/push';
+import { OrchestrationService } from '@/services/orchestration.service';
+import Container from 'typedi';
+
+type WorkflowUpdatedPayload = {
+	workflowId: string;
+	oldState: boolean;
+	newState: boolean;
+	versionId: string;
+};
+
+const isValidPayload = (payload?: Record<string, unknown>): payload is WorkflowUpdatedPayload => {
+	const { workflowId, oldState, newState, versionId } = payload ?? {};
+
+	return (
+		typeof workflowId === 'string' &&
+		typeof oldState === 'boolean' &&
+		typeof newState === 'boolean' &&
+		typeof versionId === 'string'
+	);
+};
+
+/**
+ * In multi-main setup, the leader should react to a `workflow-updated` message
+ * from a follower by managing the leader's active triggers and pollers.
+ */
+export async function handleWorkflowUpdated(payload?: Record<string, unknown>) {
+	const orchestrationService = Container.get(OrchestrationService);
+
+	if (orchestrationService.isFollower) return;
+
+	if (!isValidPayload(payload)) return;
+
+	const { workflowId, oldState, newState, versionId } = payload;
+
+	const activeWorkflowRunner = Container.get(ActiveWorkflowRunner);
+	const push = Container.get(Push);
+	const workflowRepository = Container.get(WorkflowRepository);
+
+	await activeWorkflowRunner.removeActivationError(workflowId);
+
+	if (!oldState && newState) {
+		try {
+			await activeWorkflowRunner.add(workflowId, 'activate');
+			push.broadcast('workflowActivated', { workflowId });
+
+			return;
+		} catch (error) {
+			if (error instanceof Error) {
+				await workflowRepository.update(workflowId, { active: false, versionId });
+
+				await orchestrationService.publish('workflow-failed-to-activate', {
+					workflowId,
+					errorMessage: error.message,
+				});
+			}
+
+			return;
+		}
+	}
+
+	if (oldState && !newState) {
+		await activeWorkflowRunner.remove(workflowId);
```

Comment:
One problem of this approach is the fact that the `remove` function calls `clearWebhooks` which loads the workflow from db, finds all webhooks, and then removes them.

But since the workflow has been updated already, let's say a node such as `Github trigger` was now removed. The `clearWebhooks` would not contain this node and therefore would never deregister it.

Reply:
This would mean we'd have to handle triggers differently I guess.

- Deregistration and subsequent registration of webhooks must be handled by the node receiving the request (regardless whether it's leader or not)
- Management of triggers and pollers should always be exclusively handled by the current leader, since they are always in memory and the respective `close` functions are registered there

Is this too crazy?